---
title: "Lab 7"
author: "Arianna Kazemi"
date: "3/4/2020"
output: html_document
---

```{r setup, include=FALSE}
library("BiocStyle")
library("knitr")
library("rmarkdown")
library("airway")
```

Find the pathway to the external data from the airway package. List the files in this directory and in the "quants" directory.
```{r}
dir <- system.file("extdata", package="airway", mustWork=TRUE)
list.files(dir)
list.files(file.path(dir, "quants"))
```

Create a csv with detailed information from each of the samples.
```{r}
csvfile <- file.path(dir, "sample_table.csv")
coldata <- read.csv(csvfile, row.names=1, stringsAsFactors=FALSE)
coldata
```

Create two new columns, names and files.
```{r}
coldata <- coldata[1:2,]
coldata$names <- coldata$Run
coldata$files <- file.path(dir, "quants", coldata$names, "quant.sf.gz")
file.exists(coldata$files)
```

Load tximeta and run the main function to assemble counts.
```{r}
library("tximeta")
se <- tximeta(coldata)
```

Check dimensions of dataframe and see the rownames.
```{r}
dim(se)
head(rownames(se))
```

Summarize at gene level.
```{r}
gse <- summarizeToGene(se)
```

Check that rownames are gene IDs and the dimensions are reduced.
```{r}
dim(gse)
head(rownames(gse))
```

Summary of dataframe
```{r}
par(mar=c(0,0,0,0))
plot(1,1,xlim=c(0,100),ylim=c(0,100),bty="n",
     type="n",xlab="",ylab="",xaxt="n",yaxt="n")
polygon(c(45,90,90,45),c(5,5,70,70),col="pink",border=NA)
polygon(c(45,90,90,45),c(68,68,70,70),col="pink3",border=NA)
text(67.5,40,"assay(s)")
text(67.5,35,'e.g. "counts", ...')
polygon(c(10,40,40,10),c(5,5,70,70),col="skyblue",border=NA)
polygon(c(10,40,40,10),c(68,68,70,70),col="skyblue3",border=NA)
text(25,40,"rowRanges")
polygon(c(45,90,90,45),c(75,75,95,95),col="palegreen",border=NA)
polygon(c(45,47,47,45),c(75,75,95,95),col="palegreen3",border=NA)
text(67.5,85,"colData")
```

Load whole count matrix
```{r}
data(gse)
gse
```

Look at counts in first matrix
```{r}
assayNames(gse)
head(assay(gse), 3)
colSums(assay(gse))
```

Ranges for first five and last five genes
```{r}
rowRanges(gse)
```

Chromosome metadata
```{r}
seqinfo(rowRanges(gse))
```

Look at data frame and conditions
```{r}
colData(gse)
```

Look at specific columns of gse
```{r}
gse$donor
gse$condition
```

Rename variables: donor = cell and condition = dex
```{r}
gse$cell <- gse$donor
gse$dex <- gse$condition
```

Rename levels: untreated=untrt and dexamethasone=trt
```{r}
levels(gse$dex)
# when renaming levels, the order must be preserved!
levels(gse$dex) <- c("untrt", "trt")
```

Adjust which variable is the reference/control variable
```{r}
library("magrittr")
gse$dex %<>% relevel("untrt")
gse$dex
```

Round fragments
```{r}
round( colSums(assay(gse)) / 1e6, 1 )
```

Create analysis design (accounting for cell line, what is the effect of dex)
```{r}
library("DESeq2")
dds <- DESeqDataSet(gse, design = ~ cell + dex)
```

See fragment counts
```{r}
countdata <- round(assays(gse)[["counts"]])
head(countdata, 3)
```

Save column data
```{r}
coldata <- colData(gse)
```

countdata: a table with the fragment counts
coldata: a table with information about the samples

Construct DESeqDataSet
```{r}
ddsMat <- DESeqDataSetFromMatrix(countData = countdata,
                                 colData = coldata,
                                 design = ~ cell + dex)
```

Remove rows with minimal data to speed up functions
```{r}
nrow(dds)
keep <- rowSums(counts(dds)) > 1
dds <- dds[keep,]
nrow(dds)
```

Keep at least 3 samples with a count of 10 or higher
```{r}
keep <- rowSums(counts(dds) >= 10) >= 3
```

Plot SD of each gene against the mean
```{r}
lambda <- 10^seq(from = -1, to = 2, length = 1000)
cts <- matrix(rpois(1000*100, lambda), ncol = 100)
library("vsn")
meanSdPlot(cts, ranks = FALSE)
```

Log transformed counts
```{r}
log.cts.one <- log2(cts + 1)
meanSdPlot(log.cts.one, ranks = FALSE)
```
Amplifies differences when values are close to 0 (because of +1 pseudocount).

VST: faster, less sensitive to outliers (n>30)
rlog: works well on small datasets, better for wide range of depth (n<30)

Perform VST. Counts no longer counts
```{r}
vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)
colData(vsd)
```

Perform rlog
```{r}
rld <- rlog(dds, blind = FALSE)
head(assay(rld), 3)
```
blind=false means differences bw treatment and cell lines does not contribute to expected variance-mean trend 
blind=true for unsupervised experiment 

```{r include=FALSE}
library("dplyr")
library("ggplot2")
```

Plot first sample against second. Estimate size factors to account for sequencing depth, set normalized to true.
```{r}
dds <- estimateSizeFactors(dds)

df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
         mutate(transformation = "log2(x + 1)"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"))
  
colnames(df)[1:2] <- c("x", "y")  

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation)  
```
Compress differences within low count genes, since they provide little information about differential expression.

Calculate differences between samples using VST data. Different samples should be rows, different dimensions (genes) should be columns
```{r}
sampleDists <- dist(t(assay(vsd)))
sampleDists
```

```{r}
library("pheatmap")
library("RColorBrewer")
```

Visualize differences in a heatmap. Supply distances so it does not calculate on its own.
```{r}
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$dex, vsd$cell, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```
Change row names to have treatment type and patient number 

Calculate distances using Poisson distance 
```{r}
library("PoiClaClu")
poisd <- PoissonDistance(t(counts(dds)))
```

Plot heatmap
```{r}
samplePoisDistMatrix <- as.matrix( poisd$dd )
rownames(samplePoisDistMatrix) <- paste( dds$dex, dds$cell, sep=" - " )
colnames(samplePoisDistMatrix) <- NULL
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = colors)
```

Principal components analysis (PCA): data points on 2D plane. x axis separates data points the most while y axis separates second most. There are still other dimensions of difference
```{r}
plotPCA(vsd, intgroup = c("dex", "cell"))
```

PCA plot using ggplot. Return PCA data without building plot.
```{r}
pcaData <- plotPCA(vsd, intgroup = c( "dex", "cell"), returnData = TRUE)
pcaData
percentVar <- round(100 * attr(pcaData, "percentVar"))
```

Build second plot, specifyinf cell line with symbol and treatment with color.
```{r}
ggplot(pcaData, aes(x = PC1, y = PC2, color = dex, shape = cell)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA with VST data")
```

Generalized principal component analysis (GLM-PCA) and plot
```{r}
library("glmpca")
gpca <- glmpca(counts(dds), L=2)
gpca.dat <- gpca$factors
gpca.dat$dex <- dds$dex
gpca.dat$cell <- dds$cell
ggplot(gpca.dat, aes(x = dim1, y = dim2, color = dex, shape = cell)) +
  geom_point(size =3) + coord_fixed() + ggtitle("glmpca - Generalized PCA")
```

Multidimensional scaling (MDS) for matrix of distances from VST data.
```{r}
mds <- as.data.frame(colData(vsd))  %>%
         cbind(cmdscale(sampleDistMatrix))
ggplot(mds, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with VST data")
```

Same plot with Poisson Distance
```{r}
mdsPois <- as.data.frame(colData(dds)) %>%
   cbind(cmdscale(samplePoisDistMatrix))
ggplot(mdsPois, aes(x = `1`, y = `2`, color = dex, shape = cell)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with PoissonDistances")
```



